components/server/src/workspace/env-var-service.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2023 Gitpod GmbH. All rights reserved.
+ * Licensed under the GNU Affero General Public License (AGPL).
+ * See License.AGPL.txt in the project root for license information.
+ */
+
+import { ProjectDB, UserDB } from "@gitpod/gitpod-db/lib";
+import {
+    CommitContext,
+    EnvVar,
+    EnvVarWithValue,
+    ProjectEnvVar,
+    User,
+    UserEnvVar,
+    WithEnvvarsContext,
+    Workspace,
+} from "@gitpod/gitpod-protocol";
+import { inject, injectable } from "inversify";
+import { ProjectsService } from "../projects/projects-service";
+
+export interface ResolvedEnvVars {
+    user: UserEnvVar[];
+    // all project env vars, censored included always
+    project: ProjectEnvVar[];
+    context: EnvVarWithValue[];
+    // merged workspace env vars
+    workspace: EnvVar[];
+}
+
+@injectable()
+export class EnvVarService {
+    @inject(UserDB)
+    private userDB: UserDB;
+
+    @inject(ProjectsService)
+    private projectsService: ProjectsService;
+
+    @inject(ProjectDB)
+    private projectDB: ProjectDB;
+
+    async resolve({
+        user,
+        workspace,
+        includeCensored,
+    }: {
+        user: User | undefined;
+        workspace: Workspace;
+        includeCensored: boolean;
+    }): Promise<ResolvedEnvVars> {
+        const workspaceEnvVars = new Map<String, EnvVar>();
+        const merge = (envs: EnvVar[]) => {
+            for (const env of envs) {
+                workspaceEnvVars.set(env.name, env);
+            }
+        };
+
+        // 1. first merge user envs
+        let userEnvVars = user ? await this.userDB.getEnvVars(user.id) : [];
+        if (CommitContext.is(workspace.context)) {
+            // this is a commit context, thus we can filter the env vars
+            userEnvVars = UserEnvVar.filter(
+                userEnvVars,
+                workspace.context.repository.owner,
+                workspace.context.repository.name,
+            );
+        }
```

Comment:
I wonder why do we return all user envs otherwise? Should not we have non user env vars outside of commit context?

Reply:
The commit context gives us the repository name which we use for the filtering. I.e. without the commit context we don't know the repository.

Reply:
yeah, i mean if context is not commit then user envs should be empty? it seems right now it is everything on main?